drivers/can/can_sja1000.c

```diff
@@ -0,0 +1,707 @@
+	if ((sr & CAN_SJA1000_SR_TCS) == 0) {
+		status = -EIO;
+	}
+
+	can_sja1000_tx_done(dev, status);
+}
+
+static void can_sja1000_handle_error_warning_irq(const struct device *dev)
+{
+	struct can_sja1000_data *data = dev->data;
+	uint8_t sr;
+	int err;
+
+	sr = can_sja1000_read_reg(dev, CAN_SJA1000_SR);
+	if ((sr & CAN_SJA1000_SR_BS) != 0) {
+		data->state = CAN_BUS_OFF;
+		can_sja1000_tx_done(dev, -ENETDOWN);
+#ifdef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+		/* Recover bus now unless interrupted in the middle of a MOD register change. */
+		err = k_mutex_lock(&data->mod_lock, K_NO_WAIT);
+		if (err == 0) {
+			(void)can_sja1000_leave_reset_mode(dev);
+			k_mutex_unlock(&data->mod_lock);
+		}
+#endif /* CONFIG_CAN_AUTO_BUS_OFF_RECOVERY */
+	} else if ((sr & CAN_SJA1000_SR_ES) != 0) {
+		data->state = CAN_ERROR_WARNING;
+	} else {
+		data->state = CAN_ERROR_ACTIVE;
+	}
+}
+
+static void can_sja1000_handle_error_passive_irq(const struct device *dev)
+{
+	struct can_sja1000_data *data = dev->data;
+
+	if (data->state == CAN_ERROR_PASSIVE) {
+		data->state = CAN_ERROR_WARNING;
+	} else {
+		data->state = CAN_ERROR_PASSIVE;
+	}
+}
+
+void can_sja1000_isr(const struct device *dev)
+{
+	struct can_sja1000_data *data = dev->data;
+	const can_state_change_callback_t cb = data->state_change_cb;
+	void *cb_data = data->state_change_cb_data;
+	enum can_state prev_state = data->state;
+	struct can_bus_err_cnt err_cnt;
+	uint8_t ir;
+
+	ir = can_sja1000_read_reg(dev, CAN_SJA1000_IR);
+
+	if ((ir & CAN_SJA1000_IR_TI) != 0) {
+		can_sja1000_handle_transmit_irq(dev);
+	}
+
+	if ((ir & CAN_SJA1000_IR_RI) != 0) {
+		can_sja1000_handle_receive_irq(dev);
+	}
+
+	if ((ir & CAN_SJA1000_IR_EI) != 0) {
+		can_sja1000_handle_error_warning_irq(dev);
+	}
+
+	if ((ir & CAN_SJA1000_IR_EPI) != 0) {
+		can_sja1000_handle_error_passive_irq(dev);
+	}
+
+	if (prev_state != data->state && cb != NULL) {
+		err_cnt.rx_err_cnt = can_sja1000_read_reg(dev, CAN_SJA1000_RXERR);
+		err_cnt.tx_err_cnt = can_sja1000_read_reg(dev, CAN_SJA1000_TXERR);
+		cb(dev, data->state, err_cnt, cb_data);
+	}
+}
+
+int can_sja1000_init(const struct device *dev)
+{
+	const struct can_sja1000_config *config = dev->config;
+	struct can_sja1000_data *data = dev->data;
+	struct can_timing timing;
+	int err;
+
+	__ASSERT_NO_MSG(config->read_reg != NULL);
+	__ASSERT_NO_MSG(config->write_reg != NULL);
+
+	if (config->phy != NULL) {
+		if (!device_is_ready(config->phy)) {
+			LOG_ERR("CAN transceiver not ready");
+			return -ENODEV;
+		}
+	}
+
+	k_mutex_init(&data->mod_lock);
+	k_sem_init(&data->tx_idle, 1, 1);
+	k_sem_init(&data->tx_done, 0, 1);
+
+	data->state = CAN_ERROR_ACTIVE;
+
+	/* See NXP SJA1000 Application Note AN97076 (figure 12) for initialization sequence */
+
+	/* Enter reset mode */
+	err = can_sja1000_enter_reset_mode(dev);
+	if (err != 0) {
+		return err;
+	}
+
+	/* Set PeliCAN mode */
+	can_sja1000_write_reg(dev, CAN_SJA1000_CDR, config->cdr | CAN_SJA1000_CDR_CAN_MODE);
```

Comment:
That's really unfortunate that they are not fully compatible. I was about to test the PR on an ESP32-C3, assuming it would only require to add the dts node with the correct register offset.

Reply:
Yes. It is unfortunate. I too discovered these changes between the ESP32 TWAI and those present in other Espressif SoCs.

My plan is handle these by replacing the affected generic SJA1000 API functions with custom ones for the SoCs that do not adhere to the SJA1000 register layout (e.g. using a custom init function, custom timing function). This will allow us to keep non-SJA1000 compliant functionality/extension out of the common backend and isolate it to the given front-end driver.

Reply:
Put in another way, the TWAI module present in the ESP32, ESP32-C3 and others are not the same device - and will require a different DTS `compatible` along with a new front-end driver. Most of the shared SJA1000 functionality still applies, though.